Fix LightningCLI saving hyperparameters breaking change

mauvilsa opened the pull request:
## What does this PR do?

<!--
Please include a summary of the change and which issue is fixed.
Please also include relevant motivation and context.
List any dependencies that are required for this change.

If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

The following links the related issue to the PR (https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
-->

Changes the `LightningCLI`'s `load_from_checkpoint` support such that the hparams are not saved with the `class_path` and `init_args` structure, since this constitutes a breaking change which we failed to notice in #18105.

Fixes #19977

<!-- Does your PR introduce any breaking changes? If yes, please list them. -->

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

<!-- In the CHANGELOG, separate each item in the unreleased section by a blank line to reduce collisions -->

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

<!--

Did you have fun?

Make sure you had fun coding 🙃

-->


<!-- readthedocs-preview pytorch-lightning start -->
----
📚 Documentation preview 📚: https://pytorch-lightning--20068.org.readthedocs.build/en/20068/

<!-- readthedocs-preview pytorch-lightning end -->

adamjstewart reviewed (changes requested):
Thanks for the quick fix @mauvilsa. ~This seems very specific to `init_args` and ignores `dict_kwargs`.~ I also seem to be missing many class parameters with default values that are not listed in the YAML file, and/or optional values that default to None.

EDIT: It's possible no special handling is needed for `dict_kwargs`.

adamjstewart commented:
Steps to reproduce are identical to https://github.com/Lightning-AI/pytorch-lightning/issues/19977#issuecomment-2214014470, but replace `model.py` with:
```python
from lightning.pytorch import LightningModule


class Model(LightningModule):
    def __init__(self, learning_rate, step_size=None):
        super().__init__()
        self.save_hyperparameters()
        assert 'learning_rate' in self.hparams
        assert 'step_size' in self.hparams

    def training_step(*args):
        pass

    def configure_optimizers(*args):
        pass
```
The code will fail because `step_size` is not defined in `self.hparams`.

mauvilsa commented:
> I also seem to be missing many class parameters with default values that are not listed in the YAML file, and/or optional values that default to None.

Indeed, there was a bug.

> It's possible no special handling is needed for dict_kwargs.

It is needed. It is now fixed and I added a new unit test based on your code snippet which includes `dict_kwargs`.

@adamjstewart please check again.

adamjstewart reviewed (approved):
Passes all TorchGeo tests now, thanks for the quick fix!

awaelchli reviewed (approved):
Thank you @mauvilsa, very appreciated!

mauvilsa commented on src/lightning/pytorch/cli.py:
```diff
@@ -791,17 +791,27 @@ def __init__(self, cli: LightningCLI, key: str) -> None:
+        hparams = self.cli.config_dump.get(self.key, {})
+        if "class_path" in hparams:
+            hparams = {
+                "_class_path": hparams["class_path"],
```
Added a comment.

awaelchli commented:
Thanks @mauvilsa @adamjstewart for the help here 🚀
